Add support for RTT calculation, as well as flush_timeout

spacejam opened the pull request:
I've also implemented flush_timeout to have parity with the Go client.

derekcollison commented on src/lib.rs:
````diff
@@ -842,6 +842,30 @@ impl Connection {
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn flush_timeout(&self, duration: Duration) -> io::Result<()> {
````
We should have a pain flush as well with default timeout like the go client.

Also, if the client gets disconnected it should release all pending flush calls with a disconnected error etc. Check Go client.

derekcollison commented on src/lib.rs:
```diff
@@ -821,6 +821,9 @@ impl Connection {
     }
 
     /// Flush a NATS connection by sending a `PING` protocol and waiting for the responding `PONG`.
+    /// Will fail with `TimedOut` if the server does not respond with in 60 seconds.
```
Nothing should take 60 secs, even around the world for ping I would wait max 2-5 secs.. ;)

spacejam replied:
Ok, I copied the go client. I'll change it to 3 seconds.

derekcollison replied:
lol yea, understand.

Do we unblock them in Go client on disconnect immediately or wait through reconnect attempts. If immediately let's do 5 secs and I will make the Go client that as well.

Maybe send an email to have all clients consistent.

derekcollison commented on src/lib.rs:
```diff
@@ -821,6 +821,9 @@ impl Connection {
     /// Flush a NATS connection by sending a `PING` protocol and waiting for the responding `PONG`.
+    /// Will fail with `TimedOut` if the server does not respond with in 60 seconds.
+    /// Will fail with `NotConnected` if the server is not currently connected.
+    /// Will fail with `BrokenPipe` if the connection to the server is lost.
```
Maybe make this disconnected. We may not be using sockets in the future etc.

derekcollison replied:
Meaning brokenpipe may not be accurate later on.

spacejam commented:
@derekcollison Just added a default timeout of 60 seconds to the `flush` command. Now both `flush` and `flush_timeout` will fail with `BrokenPipe` if the connection to the server is lost after beginning this request. They will fail with `NotConnected` if the server is not connected.

derekcollison commented on src/lib.rs:
```diff
@@ -831,19 +834,13 @@ impl Connection {
-        self.shared_state.outbound.send_ping()?;
-        r.recv().unwrap();
-        Ok(())
+        self.flush_timeout(Duration::from_secs(60))
```
Make this like 10, I will update Go client as well.

derekcollison commented on src/lib.rs:
```diff
@@ -831,19 +834,13 @@ impl Connection {
     /// Flush a NATS connection by sending a `PING` protocol and waiting for the responding `PONG`.
     /// Will fail with `TimedOut` if the server takes longer than this duration to respond.
+    /// Will fail with `NotConnected` if the server is not currently connected.
+    /// Will fail with `BrokenPipe` if the connection to the server is lost.
```
Same as above.

spacejam commented:
Ok, I've updated the timeout and changed the error to ConnectionReset (the closest in name to Disconnected without going down the rabbit hole + user friction of creating our own error types).

derekcollison commented on src/lib.rs:
```diff
@@ -823,7 +823,7 @@ impl Connection {
     /// Will fail with `TimedOut` if the server does not respond with in 10 seconds.
     /// Will fail with `NotConnected` if the server is not currently connected.
-    /// Will fail with `BrokenPipe` if the connection to the server is lost.
+    /// Will fail with `ConnectionReset` if the connection to the server is lost.
```
nit but we do not know yet if it has been reset, we just know we got disconnected at this point. Other client use disconnect(ed) as an event and state type.

spacejam replied:
we're using io::Error instead of creating our own. my opinion is that it really reduces user friction (especially for network-related issues) to just use the std error type around this. https://github.com/nats-io/nats.rs/pull/29#issuecomment-620035458

spacejam replied:
Disconnected is not an option among those: https://doc.rust-lang.org/std/io/enum.ErrorKind.html

I don't have a preference for which one gets used, but I do think it should be one of these rather than forcing users to handle our own error type.

derekcollison replied:
I agree with using standard errors, so fair point. That is why I coded it that way and did not create our own. Thx.

derekcollison replied:
BrokenPipe is sent back or ConnectReset? Which means by peer IIRC.

spacejam replied:
I chose BrokenPipe because that's what happens when it's not necessarily the result of a rst/fin. ConnectionReset is a rst. Which would you prefer from https://doc.rust-lang.org/std/io/enum.ErrorKind.html?

derekcollison replied:
Let's go with your original one, thanks for the patience.

spacejam commented:
Thanks for the review!
